### ISSUE TITLE
🐛 [cloud_firestore] [FATAL] Hundreds of EXC_BAD_ACCESS KERN_PROTECTION_FAILURE / KERN_INVALID_ADDRESS crashes on iOS

### ISSUE THREAD

**Issue by talent-apps** [FirebaseExtended/flutterfire#7177]
After upgrading `Firestore` version on `Podfile.lock` from `Firebase/Firestore (= 8.3.0)` to `Firebase/Firestore (= 8.6.0)`
We started seeing **hundreds** of daily iOS crashes on Crashlytics. No firestore related dart code changes were made.
The top crashes are:

```
Crashed: com.google.firebase.firestore
EXC_BAD_ACCESS KERN_PROTECTION_FAILURE 0x000000113f800020
=============================================
Crashed: com.google.firebase.firestore
0  FirebaseFirestore              0xa4350 firebase::firestore::local::LevelDbRemoteDocumentCache::GetAllExisting(firebase::firestore::immutable::SortedSet<firebase::firestore::model::DocumentKey, firebase::firestore::util::Comparator<firebase::firestore::model::DocumentKey> > const&) + 104
1  FirebaseFirestore              0xa48a0 firebase::firestore::local::LevelDbRemoteDocumentCache::GetMatching(firebase::firestore::core::Query const&, firebase::firestore::model::SnapshotVersion const&) + 496
2  FirebaseFirestore              0xaf498 firebase::firestore::local::LocalDocumentsView::GetDocumentsMatchingCollectionQuery(firebase::firestore::core::Query const&, firebase::firestore::model::SnapshotVersion const&) + 80
3  FirebaseFirestore              0xe4bf0 firebase::firestore::local::QueryEngine::GetDocumentsMatchingQuery(firebase::firestore::core::Query const&, firebase::firestore::model::SnapshotVersion const&, firebase::firestore::immutable::SortedSet<firebase::firestore::model::DocumentKey, firebase::firestore::util::Comparator<firebase::firestore::model::DocumentKey> > const&) + 404
4  FirebaseFirestore              0xbc0f0 std::__1::__function::__func<std::__1::enable_if<!(std::is_same<void, decltype(fp0())>::value), decltype(fp0())>::type firebase::firestore::local::Persistence::Run<firebase::firestore::local::LocalStore::ExecuteQuery(firebase::firestore::core::Query const&, bool)::$_14>(absl::lts_2020_02_25::string_view, firebase::firestore::local::LocalStore::ExecuteQuery(firebase::firestore::core::Query const&, bool)::$_14)::'lambda'(), std::__1::allocator<std::__1::enable_if<!(std::is_same<void, decltype(fp0())>::value), decltype(fp0())>::type firebase::firestore::local::Persistence::Run<firebase::firestore::local::LocalStore::ExecuteQuery(firebase::firestore::core::Query const&, bool)::$_14>(absl::lts_2020_02_25::string_view, firebase::firestore::local::LocalStore::ExecuteQuery(firebase::firestore::core::Query const&, bool)::$_14)::'lambda'()>, void ()>::operator()() + 356
5  FirebaseFirestore              0xa28b8 firebase::firestore::local::LevelDbPersistence::RunInternal(absl::lts_2020_02_25::string_view, std::__1::function<void ()>) + 136
6  FirebaseFirestore              0xb5614 firebase::firestore::local::LocalStore::ExecuteQuery(firebase::firestore::core::Query const&, bool) + 152
7  FirebaseFirestore              0x107bfc firebase::firestore::core::SyncEngine::InitializeViewAndComputeSnapshot(firebase::firestore::core::Query const&, int) + 84
8  FirebaseFirestore              0x1077b4 firebase::firestore::core::SyncEngine::Listen(firebase::firestore::core::Query) + 476
9  FirebaseFirestore              0x419e8 firebase::firestore::core::EventManager::AddQueryListener(std::__1::shared_ptr<firebase::firestore::core::QueryListener>) + 1096
10 FirebaseFirestore              0x646b0 std::__1::__function::__func<firebase::firestore::core::FirestoreClient::ListenToQuery(firebase::firestore::core::Query, firebase::firestore::core::ListenOptions, std::__1::shared_ptr<firebase::firestore::core::EventListener<firebase::firestore::core::ViewSnapshot> >&&)::$_10, std::__1::allocator<firebase::firestore::core::FirestoreClient::ListenToQuery(firebase::firestore::core::Query, firebase::firestore::core::ListenOptions, std::__1::shared_ptr<firebase::firestore::core::EventListener<firebase::firestore::core::ViewSnapshot> >&&)::$_10>, void ()>::operator()() + 68
11 FirebaseFirestore              0x6614 firebase::firestore::util::AsyncQueue::ExecuteBlocking(std::__1::function<void ()> const&) + 68
12 FirebaseFirestore              0x114a08 firebase::firestore::util::Task::ExecuteAndRelease() + 184
13 libdispatch.dylib              0x481c _dispatch_client_callout + 20
14 libdispatch.dylib              0xc004 _dispatch_lane_serial_drain + 620
15 libdispatch.dylib              0xcc00 _dispatch_lane_invoke + 404
16 libdispatch.dylib              0x174bc _dispatch_workloop_worker_thread + 764
17 libsystem_pthread.dylib        0x37a4 _pthread_wqthread + 276
18 libsystem_pthread.dylib        0xa74c start_wqthread + 8
```

```
Crashed: com.google.firebase.firestore
EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x49742400c7c35000
=============================================
Crashed: com.google.firebase.firestore
0  FirebaseFirestore              0x8a18 firebase::firestore::util::ComparisonResult firebase::firestore::util::CompareContainer<std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > >(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) + 24
1  FirebaseFirestore              0xa6748 firebase::firestore::immutable::impl::LlrbNode<firebase::firestore::model::DocumentKey, firebase::firestore::model::MutableDocument> firebase::firestore::immutable::impl::LlrbNode<firebase::firestore::model::DocumentKey, firebase::firestore::model::MutableDocument>::InnerErase<firebase::firestore::util::Comparator<firebase::firestore::model::DocumentKey> >(firebase::firestore::model::DocumentKey const&, firebase::firestore::util::Comparator<firebase::firestore::model::DocumentKey> const&) const + 112
2  FirebaseFirestore              0xa66ac firebase::firestore::immutable::impl::TreeSortedMap<firebase::firestore::model::DocumentKey, firebase::firestore::model::MutableDocument, firebase::firestore::util::Comparator<firebase::firestore::model::DocumentKey> >::erase(firebase::firestore::model::DocumentKey const&) const + 36
3  FirebaseFirestore              0xa45e0 firebase::firestore::immutable::SortedMap<firebase::firestore::model::DocumentKey, firebase::firestore::model::MutableDocument, firebase::firestore::util::Comparator<firebase::firestore::model::DocumentKey> >::erase(firebase::firestore::model::DocumentKey const&) const + 96
4  FirebaseFirestore              0xa4368 firebase::firestore::local::LevelDbRemoteDocumentCache::GetAllExisting(firebase::firestore::immutable::SortedSet<firebase::firestore::model::DocumentKey, firebase::firestore::util::Comparator<firebase::firestore::model::DocumentKey> > const&) + 128
5  FirebaseFirestore              0xa48a0 firebase::firestore::local::LevelDbRemoteDocumentCache::GetMatching(firebase::firestore::core::Query const&, firebase::firestore::model::SnapshotVersion const&) + 496
6  FirebaseFirestore              0xaf498 firebase::firestore::local::LocalDocumentsView::GetDocumentsMatchingCollectionQuery(firebase::firestore::core::Query const&, firebase::firestore::model::SnapshotVersion const&) + 80
7  FirebaseFirestore              0xe4bf0 firebase::firestore::local::QueryEngine::GetDocumentsMatchingQuery(firebase::firestore::core::Query const&, firebase::firestore::model::SnapshotVersion const&, firebase::firestore::immutable::SortedSet<firebase::firestore::model::DocumentKey, firebase::firestore::util::Comparator<firebase::firestore::model::DocumentKey> > const&) + 404
8  FirebaseFirestore              0xbc0f0 std::__1::__function::__func<std::__1::enable_if<!(std::is_same<void, decltype(fp0())>::value), decltype(fp0())>::type firebase::firestore::local::Persistence::Run<firebase::firestore::local::LocalStore::ExecuteQuery(firebase::firestore::core::Query const&, bool)::$_14>(absl::lts_2020_02_25::string_view, firebase::firestore::local::LocalStore::ExecuteQuery(firebase::firestore::core::Query const&, bool)::$_14)::'lambda'(), std::__1::allocator<std::__1::enable_if<!(std::is_same<void, decltype(fp0())>::value), decltype(fp0())>::type firebase::firestore::local::Persistence::Run<firebase::firestore::local::LocalStore::ExecuteQuery(firebase::firestore::core::Query const&, bool)::$_14>(absl::lts_2020_02_25::string_view, firebase::firestore::local::LocalStore::ExecuteQuery(firebase::firestore::core::Query const&, bool)::$_14)::'lambda'()>, void ()>::operator()() + 356
9  FirebaseFirestore              0xa28b8 firebase::firestore::local::LevelDbPersistence::RunInternal(absl::lts_2020_02_25::string_view, std::__1::function<void ()>) + 136
10 FirebaseFirestore              0xb5614 firebase::firestore::local::LocalStore::ExecuteQuery(firebase::firestore::core::Query const&, bool) + 152
11 FirebaseFirestore              0x107bfc firebase::firestore::core::SyncEngine::InitializeViewAndComputeSnapshot(firebase::firestore::core::Query const&, int) + 84
12 FirebaseFirestore              0x1077b4 firebase::firestore::core::SyncEngine::Listen(firebase::firestore::core::Query) + 476
13 FirebaseFirestore              0x419e8 firebase::firestore::core::EventManager::AddQueryListener(std::__1::shared_ptr<firebase::firestore::core::QueryListener>) + 1096
14 FirebaseFirestore              0x646b0 std::__1::__function::__func<firebase::firestore::core::FirestoreClient::ListenToQuery(firebase::firestore::core::Query, firebase::firestore::core::ListenOptions, std::__1::shared_ptr<firebase::firestore::core::EventListener<firebase::firestore::core::ViewSnapshot> >&&)::$_10, std::__1::allocator<firebase::firestore::core::FirestoreClient::ListenToQuery(firebase::firestore::core::Query, firebase::firestore::core::ListenOptions, std::__1::shared_ptr<firebase::firestore::core::EventListener<firebase::firestore::core::ViewSnapshot> >&&)::$_10>, void ()>::operator()() + 68
15 FirebaseFirestore              0x6614 firebase::firestore::util::AsyncQueue::ExecuteBlocking(std::__1::function<void ()> const&) + 68
16 FirebaseFirestore              0x114a08 firebase::firestore::util::Task::ExecuteAndRelease() + 184
17 libdispatch.dylib              0x61298 _dispatch_client_callout + 16
18 libdispatch.dylib              0x9a40 _dispatch_lane_serial_drain$VARIANT$mp + 612
19 libdispatch.dylib              0xa518 _dispatch_lane_invoke$VARIANT$mp + 420
20 libdispatch.dylib              0x13fac _dispatch_workloop_worker_thread + 712
21 libsystem_pthread.dylib        0xb5bc _pthread_wqthread + 272
22 libsystem_pthread.dylib        0xe86c start_wqthread + 8
```

```
Crashed: com.google.firebase.firestore
EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x0000000000000000
=============================================
Crashed: com.google.firebase.firestore
0  FirebaseFirestore              0xa5c60 firebase::firestore::immutable::impl::LlrbNodeIterator<firebase::firestore::immutable::impl::LlrbNode<firebase::firestore::model::DocumentKey, firebase::firestore::model::MutableDocument> >::AccumulateLeft(firebase::firestore::immutable::impl::LlrbNode<firebase::firestore::model::DocumentKey, firebase::firestore::model::MutableDocument> const*, std::__1::stack<firebase::firestore::immutable::impl::LlrbNode<firebase::firestore::model::DocumentKey, firebase::firestore::model::MutableDocument> const*, std::__1::deque<firebase::firestore::immutable::impl::LlrbNode<firebase::firestore::model::DocumentKey, firebase::firestore::model::MutableDocument> const*, std::__1::allocator<firebase::firestore::immutable::impl::LlrbNode<firebase::firestore::model::DocumentKey, firebase::firestore::model::MutableDocument> const*> > >*) + 72
1  FirebaseFirestore              0xa6410 firebase::firestore::immutable::impl::LlrbNodeIterator<firebase::firestore::immutable::impl::LlrbNode<firebase::firestore::model::DocumentKey, firebase::firestore::model::MutableDocument> >::operator++() + 84
2  FirebaseFirestore              0xa43a4 firebase::firestore::local::LevelDbRemoteDocumentCache::GetAllExisting(firebase::firestore::immutable::SortedSet<firebase::firestore::model::DocumentKey, firebase::firestore::util::Comparator<firebase::firestore::model::DocumentKey> > const&) + 188
3  FirebaseFirestore              0xa48a0 firebase::firestore::local::LevelDbRemoteDocumentCache::GetMatching(firebase::firestore::core::Query const&, firebase::firestore::model::SnapshotVersion const&) + 496
4  FirebaseFirestore              0xaf498 firebase::firestore::local::LocalDocumentsView::GetDocumentsMatchingCollectionQuery(firebase::firestore::core::Query const&, firebase::firestore::model::SnapshotVersion const&) + 80
5  FirebaseFirestore              0xe4bf0 firebase::firestore::local::QueryEngine::GetDocumentsMatchingQuery(firebase::firestore::core::Query const&, firebase::firestore::model::SnapshotVersion const&, firebase::firestore::immutable::SortedSet<firebase::firestore::model::DocumentKey, firebase::firestore::util::Comparator<firebase::firestore::model::DocumentKey> > const&) + 404
6  FirebaseFirestore              0xbc0f0 std::__1::__function::__func<std::__1::enable_if<!(std::is_same<void, decltype(fp0())>::value), decltype(fp0())>::type firebase::firestore::local::Persistence::Run<firebase::firestore::local::LocalStore::ExecuteQuery(firebase::firestore::core::Query const&, bool)::$_14>(absl::lts_2020_02_25::string_view, firebase::firestore::local::LocalStore::ExecuteQuery(firebase::firestore::core::Query const&, bool)::$_14)::'lambda'(), std::__1::allocator<std::__1::enable_if<!(std::is_same<void, decltype(fp0())>::value), decltype(fp0())>::type firebase::firestore::local::Persistence::Run<firebase::firestore::local::LocalStore::ExecuteQuery(firebase::firestore::core::Query const&, bool)::$_14>(absl::lts_2020_02_25::string_view, firebase::firestore::local::LocalStore::ExecuteQuery(firebase::firestore::core::Query const&, bool)::$_14)::'lambda'()>, void ()>::operator()() + 356
7  FirebaseFirestore              0xa28b8 firebase::firestore::local::LevelDbPersistence::RunInternal(absl::lts_2020_02_25::string_view, std::__1::function<void ()>) + 136
8  FirebaseFirestore              0xb5614 firebase::firestore::local::LocalStore::ExecuteQuery(firebase::firestore::core::Query const&, bool) + 152
9  FirebaseFirestore              0x107bfc firebase::firestore::core::SyncEngine::InitializeViewAndComputeSnapshot(firebase::firestore::core::Query const&, int) + 84
10 FirebaseFirestore              0x1077b4 firebase::firestore::core::SyncEngine::Listen(firebase::firestore::core::Query) + 476
11 FirebaseFirestore              0x419e8 firebase::firestore::core::EventManager::AddQueryListener(std::__1::shared_ptr<firebase::firestore::core::QueryListener>) + 1096
12 FirebaseFirestore              0x646b0 std::__1::__function::__func<firebase::firestore::core::FirestoreClient::ListenToQuery(firebase::firestore::core::Query, firebase::firestore::core::ListenOptions, std::__1::shared_ptr<firebase::firestore::core::EventListener<firebase::firestore::core::ViewSnapshot> >&&)::$_10, std::__1::allocator<firebase::firestore::core::FirestoreClient::ListenToQuery(firebase::firestore::core::Query, firebase::firestore::core::ListenOptions, std::__1::shared_ptr<firebase::firestore::core::EventListener<firebase::firestore::core::ViewSnapshot> >&&)::$_10>, void ()>::operator()() + 68
13 FirebaseFirestore              0x6614 firebase::firestore::util::AsyncQueue::ExecuteBlocking(std::__1::function<void ()> const&) + 68
14 FirebaseFirestore              0x114a08 firebase::firestore::util::Task::ExecuteAndRelease() + 184
15 libdispatch.dylib              0x481c _dispatch_client_callout + 20
16 libdispatch.dylib              0xc004 _dispatch_lane_serial_drain + 620
17 libdispatch.dylib              0xcc00 _dispatch_lane_invoke + 404
18 libdispatch.dylib              0x174bc _dispatch_workloop_worker_thread + 764
19 libsystem_pthread.dylib        0x37a4 _pthread_wqthread + 276
20 libsystem_pthread.dylib        0xa74c start_wqthread + 8
```

Output of `flutter doctor -v`:
```
[√] Flutter (Channel stable, 2.5.2, on Microsoft Windows [Version 10.0.19043.1237], locale en-US)
    • Flutter version 2.5.2 at C:\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 3595343e20 (12 days ago), 2021-09-30 12:58:18 -0700
    • Engine revision 6ac856380f
    • Dart version 2.14.3

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\XXX\AppData\Local\Android\Sdk
    • Platform android-31, build-tools 30.0.3
    • ANDROID_HOME = C:\Users\XXX\AppData\Local\Android\Sdk
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java      
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)    
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe     

[√] Android Studio (version 2020.3)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:

C:\XXX\Dev\FlutterProjects\winner>flutter doctor -v
[√] Flutter (Channel stable, 2.5.2, on Microsoft Windows [Version 10.0.19043.1237], locale en-US)
    • Flutter version 2.5.2 at C:\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 3595343e20 (12 days ago), 2021-09-30 12:58:18 -0700
    • Engine revision 6ac856380f
    • Dart version 2.14.3

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\XXX\AppData\Local\Android\Sdk
    • Platform android-31, build-tools 30.0.3
    • ANDROID_HOME = C:\Users\XXX\AppData\Local\Android\Sdk
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java      
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)    
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe     

[√] Android Studio (version 2020.3)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[√] VS Code (version 1.61.0)
    • VS Code at C:\Users\XXX\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.27.0

[√] Connected device (2 available)
    • Chrome (web) • chrome • web-javascript • Google Chrome 94.0.4606.81
    • Edge (web)   • edge   • web-javascript • Microsoft Edge 94.0.992.38

• No issues found!
```

Please advise. It's a total disaster ATM.

**Comment by darshankawar** [comment]
> but the `firebase_core` plugin still uses `8.7.0`

That is correct. It was bumped up in latest plugin version. You may need to wait until a new plugin version is released which may use `8.8.0` of Firebase iOS SDK.

**Comment by talent-apps** [comment]
We used a hardcoded Firebase iOS 8.8.0 version and for now it seems to solve the problem in production.
Considering the severity of this issue, it will help expedite a new version of `firebase_core` with 8.8.0.
